### PR TITLE
package upload

### PR DESCRIPTION
**Pull request by leng20000**

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -30,6 +30,92 @@ pub async fn test() -> &'static str {
...
+    package: Json<Package>,
+    mod_db: &State<ModuleDB>,
+) -> (Status, &'static str) {
+    let mod_r = mod_db.read().await;
+    let res = mod_r.get(&id);
+    if res.is_none() {
+        return (Status::NotFound, "Package does not exist.");
+    }
+
+    let db = res.unwrap();
```
Comment (starliteSeeker) [inline src/api.rs]: For upload, the package *shouldn't* exist before, so there's no way of finding an entry in `mod_db`. If the package existed, its a 409 error. I imagine that if you set the id properly (say some combination of owner and repo name), you should be able to check if the  package already existed in the registry just by checking for the id.

**Review by starliteSeeker — COMMENTED** [review]
There's no 404 for this API in the spec, so you can't return 404.

Also, you'll have to rate the package and reject it (424) if it doesn't match expectations. Either doing it in `Module::new` or checking the fields in the return value of `Module::new()` should be good (I also realize I didn't do this for package update, can you add it for me?)

You will need the URL for rating packages, so if the user provides Content and not URL, you have to look into `package.json` in the zip file the user provided. (I imagine the order would be download package to some path -> get url in `package.json` -> call `Module::new()` -> move the package to the correct path)

**Comment by leng20000** [comment]
I'm not sure if we discussed this before or if its mentioned anywhere, but what's the lowest score that we're going to accept for each module? I'm using 0.5 as of now. 

In the latest commit, I have made changes to handle the URL field. However, I'm still working on handling Content field. If there's no 404 error code in the API, then should I replace it with other error code specified in the yaml file?

**Comment by starliteSeeker** [comment]
It's written somewhere but the threashhold is 0.5 if my memory is to be trusted.

For 404, yeah replace it with other errors that's listed in the API. If you feel like none of the error messages applies, I guess just use 400.

Some other things: PackageCreate is in `/package`, so there's no id field for you to do `mod_r.get(&id);`. Also it takes in a `PackageData`, not `Package`, and returns `Package` for 201. Since the APIs are autograded (I think) there's no choice but to follow the specs very closely.

**Comment by leng20000** [comment]
Since it takes in `PackageData`, we don't have a `name` variable to call `get_by_name` to check if the package already exist or not. So this means that we have to get the name from the `Content` and `URL` first? 

But, I'm not too sure how to get the extract the name given the Content field. For URL, I would have to extract the repo name in the github/npm link right?

**Comment by starliteSeeker** [comment]
Searching in the hashmap by name is more expensive than searching by id, so I'd say search by id (`get()`) if possible.

But onto your question, if you have Contents, you can peek into `package.json` to see the package name on npm, which means that `https://www.npmjs.com/package/<name>` is the URL. Pass this URL into `extract_owner_and_repo()` to get the owner and repo name.

If you have npm URL, the package name is in the URL. Pass this URL into `extract_owner_and_repo()` to get the owner and repo name.

If you have github URL, pass it into `extract_owner_and_repo()` to get the owner and repo name.You can download the package, check `package.json` and you can find the package name

Since the package name is unique, you can probably get away with setting `id` = `name` = package name in `Module` and not deal with owner and repo name at all. Checkout `match_readme()` in `api.rs` for an example of reading a file in zip.

**Comment by leng20000** [comment]
Sorry that it's taking me more time than I originally expected, but one other question is what are the differences between the name and the repo? Shouldn't it be the same? If that is the case, then simply extracting the repo name from the URL is not enough?

**Comment by starliteSeeker** [comment]
In my mind, there's the repo name (module-registry in `https://github.com/lee3445/module-registry`) and the package name (postcss in `https://www.npmjs.com/package/postcss`, and in package.json), and I don't think repo name and package name have to be equal. For repo name, different owner can have repo with the same name, but for npm, the package name is unique, which is why I said the package name might be able to serve the purpose of id.

**Comment by starliteSeeker** [comment]
You didn't put the entry in main, please test your stuff, or at least get rid of the warnings that can be avoided.

For the one package I tested it with (https://github.com/dcousens/is-sorted), the version is not extracted from `package.json` correctly (it's "0"), but I'm not sure what the cause is yet.

**Comment by starliteSeeker** [comment]
Seems like CI is complaining about the tests inside `database.rs` because we changed the argument for `Module::new()`. Removing the first argument should work.

**Comment by leng20000** [comment]
I fixed the stuff in main and got rid of the warning as well as the CI stuff on `database.rs` file. But I'm still not sure how to test this, the command I used to send post request is 

`$ curl http://127.0.0.1:8080/package_create --request POST -d '{"Content": null, "URL": "https://github.com/dcousens/is-sorted", "JSProgram": null}' -v`

Is this command similar to the one ran on your machine?

**Comment by leng20000** [comment]
Ok never mind what I said about getting the CI fixed. I tried changing the `database.rs` but it seems like its throwing another error, maybe this is because I deleted something else that's not related? But at the same time I only deleted 1 arguments from the test which is the same as what you were saying in the comments earlier right?

File: src/api.rs [hunk 2]
```diff
@@ -27,7 +27,203 @@ pub async fn world() -> Option<NamedFile> {
...
+    if package.Content != None && package.URL == None {
+        if let Ok(_) = base64_to_zip(
+            package.Content.as_ref().unwrap().as_str(),
+            "./temp_ece461.zip",
+        )
+        .await
+        {
+            let name = get_package("./temp_ece461.zip", "name");
+            let version = get_package("./temp_ece461.zip", "version");
+            let url = format!("https://www.npmjs.com/package/{:?}", name);
```
Comment (starliteSeeker) [inline src/api.rs]: `name` has type `Option<String>`, so I don't think you want it passed into the format function. Also be careful that `{:?}` prints strings with double quotes around them, so you probably want `{}` instead.

**Review by starliteSeeker — COMMENTED** [review]
URL works well.

**Comment by starliteSeeker** [comment]
> Ok never mind what I said about getting the CI fixed. I tried changing the `database.rs` but it seems like its throwing another error, maybe this is because I deleted something else that's not related? But at the same time I only deleted 1 arguments from the test which is the same as what you were saying in the comments earlier right?

Yeah I was too naive. So it turns out I checked if the ID is correct in one of the tests. Since we now changed the function to auto-generate the ID, we have to change the test as well. Check out where it complained in the test log, and change "postcss" to "postcss_postcss", then we should be good.

**Comment by starliteSeeker** [comment]
While not wrong, you asked for the read lock but then dropped it without ever using it, which slows the process for no reason. To save time, I'm going to take the liberty of deleting the code for the read lock.

There's one bug where the code panics if the base64 is wrong, but that's related to an error in the converting part of the code I didn't catch and not your code, so I'm going to fix it then merge it.

Well done.

**Review by starliteSeeker — APPROVED** [review]

**Pull request merged**